jujusvg.go

```diff
@@ -10,11 +12,51 @@ import (
 	"gopkg.in/juju/charm.v5-unstable"
 )
 
-// NewFromBundle returns a new Canvas that can be used
+func NewFromBundleEmbedIcons(b *charm.BundleData, iconURL func(*charm.Reference) string) (*Canvas, error) {
+	alreadyFetched := make(map[string]bool)
+	icons := make(map[string]string)
+	for _, serviceData := range b.Services {
+		charmId, err := charm.ParseReference(serviceData.Charm)
+		if err != nil {
+			return nil, errgo.Notef(err, "cannot parse charm %q", serviceData.Charm)
+		}
+		if _, ok := alreadyFetched[charmId.Path()]; ok == false {
+			alreadyFetched[charmId.Path()] = true
+			icon, err := fetchIcon(iconURL(charmId))
+			if err != nil {
+				return nil, err
+			}
+			icons[charmId.Path()] = icon
+		}
+	}
+	return NewFromBundleWithMap(b, iconURL, icons)
+}
+
+func fetchIcon(url string) (string, error) {
+	resp, err := http.Get(url)
```

Comment:
In answer to your question I would probably want to avoid the library fetching directly without allowing any overriding of settings. I think it would be good if the library user provided a fetcher function of interface that would load the icon given the charmId, that way it can do clever things with resolving, you might consider that where you put in links too.

If you are going to fetch over http then it's polite to make some way to override the http.Client that is being used (falling back to the default) to allows cookies and such to be configured.